### PR TITLE
Update TE v2 executor tests

### PR DESCRIPTION
**Pull request by riccardofelluga**
While working on #2344 I've noticed that when using mse we cannot do more than 3 iterations otherwise the loss goes outside well fp8 range. Using cross entropy helps testing a more realistic scenario and leaves more iterations for numerical errors to accumulate in case there is a mismatch with TE behavior

### CODE REVIEW

File: thunder/tests/test_transformer_engine_v2_executor.py [hunk 1]
```diff
@@ -126,16 +126,15 @@ def clone_params(*params):
...
 
     inputs = tuple(torch.rand(*input_shape, device=device, dtype=dtype) for _ in range(iterations))
 
     def train_model(model, optimizer):
         # Run for `iterations`.
         for iter_n in range(iterations):
             x = inputs[iter_n]
-            result = model(x)
-            loss = torch.nn.functional.mse_loss(result.sum(), target_value)
+            loss = torch.nn.functional.cross_entropy(result, target_value)
```
Comment (kshitij12345) [inline thunder/tests/test_transformer_engine_v2_executor.py:138]:
```suggestion
            result = model(x)
            loss = torch.nn.functional.cross_entropy(result, target_value)
```

File: thunder/executors/transformer_engine_v2ex_impl.py [hunk 2]
```diff
@@ -140,6 +140,7 @@ def _linear_fwd_impl(a, w, bias, input_quantizer: Quantizer, weight_quantizer: Q
         input=a,
         weight=w,
         bias=bias,
+        dtype=w.dtype,
```
Comment (kshitij12345) [inline thunder/executors/transformer_engine_v2ex_impl.py]: Is this necessary or is this a stray change?

Comment (riccardofelluga) [reply thunder/executors/transformer_engine_v2ex_impl.py]: I wanted to add it in the checkpointing PR but it is needed also here for new versions of TE since they changed the input validation for the _functional_forward call. So it can go up with this or the other PR

Comment (kshitij12345) [reply thunder/executors/transformer_engine_v2ex_impl.py]: It looks like a bug in TE.

https://github.com/NVIDIA/TransformerEngine/blob/858755c0c3898d5bc971fcec031b26ff26e7e4fb/transformer_engine/pytorch/ops/basic/basic_linear.py#L404-L408

`elif` should be `elif weight is not None and isinstance(weight, torch.Tensor):`.

Will file a bug in TE.

Comment (kshitij12345) [reply thunder/executors/transformer_engine_v2ex_impl.py]: Filed https://github.com/NVIDIA/TransformerEngine/issues/2011. I think we should add a comment referring to this issue above this line.

**Comment by riccardofelluga** [comment]
> LGTM, thanks @riccardofelluga
> 
> I will send a similar patch for TE v1 executor test.

I can add it with this PR if it's ok @kshitij12345

**Comment by kshitij12345** [comment]
> I can add it with this PR if it's ok @kshitij12345

Yes, that'd be great, thanks! 🙇

**Comment by riccardofelluga** [comment]
Updated and tested on both H100 and B200.

**Review by t-vi — APPROVED** [review]
Thank you @riccardofelluga @kshitij12345
